### ISSUE TITLE
plone.app.redirector support for Volto

### ISSUE THREAD

**Issue by tisto** [plone/volto#1095]
It would be nice to have plone.app.redirector support for Volto.

plone.restapi currently returns the content object itself on a redirect path.

Say you have a page "my-page" and you rename it to "my-new-page", then plone.restapi will return the same content object for:

/my-page -> 200 OK

and

/my-new-page -> 200 OK

Volto currently has no way to detect that "/my-new-page" is a redirect.

plone.restapi could return a 301 instead:

/my-new-page -> 301 Redirect

Though, Volto would need to being able to handle this. Right now, as expected Volto show the content of the original object. Which might be fine for the user but causes a problem for search engines who will detect lots of duplicated content.

I am happy to work on the plone.restapi part of this but I need a Volto master like @robgietema or @sneridagh to help me with the Volto part. :)

I would also be fine if Volto would just show a page "This link is outdated, please click on <REDIRECT_LINK>". We do not need a super-sophisticated solution as a first step.

**Comment by cekk** [comment]
It looks like restapi correctly respond with a 301 if you try to call a get with the old url on a renamed object..or am i wrong?

Probably there should be something to fix in Volto right now, because it doesn't seems to work properly: i get and error if i try to access the old url from volto. And there are some calls blocked by cors.

Does anyone started working/looking on this?

**Comment by ale-rt** [comment]
Note that the proper codes are 302 and 307, see https://github.com/plone/plone.app.redirector/issues/8 and https://github.com/plone/plone.app.redirector/pull/22

**Comment by tisto** [comment]
@cekk we started to look into this for a project and it kind of basically works. Though, there are still edge cases to fix and test in production. I am +1 on changing plone.restapi to return 302 instead of 301.

**Comment by cekk** [comment]
ok, i tried last version of p.a.redirector but the api still return a 301..maybe there is still something to do in restapi?

@tisto is there something to see for Volto somewhere? We also need to start looking into it and don't want to do the fix twice if you already started it

**Comment by tisto** [comment]
@cekk if I am not mistaken we used Volto out-of-the-box and just added some redirects and it "kind-of-worked" but not as reliable and consistent as we hoped. Also the current classic p.a.redirector controlpanel did not work at all for us. Therefore we dropped the idea for the time being. Though, I'd be happy to pick it up again and discuss how we can accomplish this.

**Comment by cekk** [comment]
@tisto if you have time, this week i'd like to discuss about that

**Comment by tisto** [comment]
plone.rest indeed returns a 301 properly: 

https://github.com/plone/plone.rest/blob/3886a5409ed93f195350f3935a2af5e5ba69f5bb/src/plone/rest/tests/test_redirects.py#L31

**Comment by sneridagh** [comment]
I gave it a quick try and it seems that superagent is unable to catch the 301 in the browsers, since it’s a browser limitation.

https://github.com/visionmedia/superagent/issues/1385

I will continue investigating :( although I don’t see how we are going to do that coming from the app (navigating, not SSR). Maybe marking the response with a header, then act?

**Comment by tiberiuichim** [comment]
I remember an idea that I've expressed at one point on this discussions: why should Volto care about the redirector? When fetching content, if that content comes from a different URL then expected, then the window location should just be updated to match the incoming `@id` for the content.

**Comment by tisto** [comment]
@cekk any update on the 301 redirection issue from your side? We ran into this again since a client of ours is doing SEO optimizations. I guess we have to solve this problem at some point because otherwise Google with down rank any pages in Plone/Volto that are moved. I had a brief discussion with @sneridagh about it to refresh the status quo. Would you be up for another call discussing the options to work around or solve this problem?

**Comment by cekk** [comment]
I wanted to talk with @sneridagh  too some weeks ago because we are in your same situation, but I didn't had the time  ;)

Yes, I absolutely want to discuss about that and fix it..I tried something but without luck.

When you want, I'm in

**Comment by sneridagh** [comment]
@tisto @cekk we finally solved this by acting here and there (I think @mamico and @cekk), right? I just tried in https://6.demo.plone.org/ and works like a charm, so a 301 happens.

Internally it does work well too, with the exception of #3702 and the redirect control panel (aliases) is also merged an in place. So I'd close this one. What do you think?

**Comment by erral** [comment]
> @tisto @cekk we finally solved this by acting here and there (I think @mamico and @cekk), right? I just tried in https://6.demo.plone.org/ and works like a charm, so a 301 happens.
> 

Although plone.restapi returns an HTTP 301 response, what Volto does is not a real HTTP 301 response, it is a url change that happens in the browser.

This happens also in the Link view see #3302 

I think we can close this issue, but we need to further investigate issuing proper HTTP 301/302/whatever response codes from Volto in such cases.
